### DISCUSSION TITLE
NextResponse not setting cookie

### DISCUSSION THREAD

**Discussion by anstapol** [vercel/community#204] [Show & Tell]
Using Next.js middleware feature I want to assign a cookie for all pages that are within a directory.

Directory structure:
```
pages/
   onboard/
      _middleware.ts
      index.tsx
      [id].tsx
```

**_middleware.ts**
```ts
export default async function middleware(req: NextRequest) {
  const { page: { name } } = req;
  const res = NextResponse.next();

  if (name) {
    console.log('Should check cookies');

    if (!req.cookies['uuid']) {
      const uuid = uuid4();

      res.cookie('uuid', uuid);

      console.log('Uuid: ', uuid);
    }

    if (!req.cookies['ab.onboard']) {
      const bucket = pickBucket('onboard');

      res.cookie('ab.onboard', bucket);

      console.log('Bucket: ', bucket);
    }

    return res;
  }

  return res;
}
```

**Middleware log**
<img width="1022" alt="middleware" src="https://user-images.githubusercontent.com/33395021/150084370-b04da279-50e3-4bec-ae35-e91168359600.png">

But the browser cookies are empty...

Only when I visit `/onboard/1` I get the `uuid` cookie and navigating further I get the `ab.onnoard` one.

--- 
**By the way, those pages are Server Side Generated.**

**Comment by Pieparker** [comment]
Hey there! It appears as though all routes are triggering the middleware but there's perhaps something wrong in the code (something to do with the nested `IF`s perhaps?). To help you debug, it would be worth stripping it back to a minimal reproduction to consider only setting the `ab.onnoard` cookie first.

With this in mind, this question may be better suited for the [Next.js GitHub Discussions](https://github.com/vercel/next.js/discussions)!
